src/main/telemetry/esc_telemetry.c

```diff
@@ -0,0 +1,330 @@
+
+0: current motor index requested
+1: number of timeouts
+2: voltage
+3: current
+*/
+
+typedef struct {
+    bool skipped;
+    uint8_t temperature;
+    uint16_t voltage;
+    uint16_t current;
+    uint16_t consumption;
+    uint16_t rpm;
+} esc_telemetry_t;
+
+typedef enum {
+    ESC_TLM_FRAME_PENDING = 1 << 0,     // 1
+    ESC_TLM_FRAME_COMPLETE = 1 << 1     // 2
+} escTlmFrameState_t;
+
+typedef enum {
+    ESC_TLM_TRIGGER_WAIT = 0,
+    ESC_TLM_TRIGGER_READY = 1 << 0,     // 1
+    ESC_TLM_TRIGGER_PENDING = 1 << 1,   // 2
+} escTlmTriggerState_t;
+
+#define ESC_TLM_BAUDRATE 115200
+#define ESC_TLM_BUFFSIZE 10
+#define ESC_BOOTTIME 3000               // 3 seconds
+#define ESC_REQUEST_TIMEOUT 1000        // 1 seconds
+
+static bool tlmFrameDone = false;
+static bool firstCycleComplete = false;
+static uint8_t tlm[ESC_TLM_BUFFSIZE] = { 0, };
+static uint8_t tlmFramePosition = 0;
+static serialPort_t *escTelemetryPort = NULL;
+static esc_telemetry_t escTelemetryData[4];
+static uint32_t escTriggerTimestamp = -1;
+
+static uint8_t escTelemetryMotor = 99;      // motor index 0 - 3
+static bool escTelemetryEnabled = false;
+static escTlmTriggerState_t escTelemetryTriggerState = ESC_TLM_TRIGGER_WAIT;
+
+static uint16_t escVbat = 0;
+static uint16_t escCurrent = 0;
+static uint16_t escConsumption = 0;
+
+static void escTelemetryDataReceive(uint16_t c);
+static uint8_t update_crc8(uint8_t crc, uint8_t crc_seed);
+static uint8_t get_crc8(uint8_t *Buf, uint8_t BufLen);
+static void selectNextMotor(void);
+
+static motorDmaOutput_t dmaMotors[MAX_SUPPORTED_MOTORS];
+
+bool isEscTelemetryActive(void)
+{
+    return escTelemetryEnabled;
+}
+
+uint16_t getEscTelemetryVbat(void)
+{
+    return escVbat / 10;
+}
+
+uint16_t getEscTelemetryCurrent(void)
+{
+    return escCurrent;
+}
+
+uint16_t getEscTelemetryConsumption(void)
+{
+    return escConsumption;
+}
+
+void pwmRequestTelemetry(uint8_t index)
+{
+    if (escTelemetryTriggerState == ESC_TLM_TRIGGER_PENDING)
+    {
+        return;
+    }
+
+    if (escTelemetryTriggerState == ESC_TLM_TRIGGER_READY)
+    {
+        if (debugMode == DEBUG_ESC_TELEMETRY) debug[0] = escTelemetryMotor+1;
+
+        if (escTelemetryMotor == index) {
+            escTelemetryTriggerState = ESC_TLM_TRIGGER_PENDING;
+            motorDmaOutput_t * const motor = &dmaMotors[index];
+            motor->requestTelemetry = true;
+        }
+    }
+}
+
+// bool escTelemetrySendTrigger(uint8_t index)
+// {
+//     if (escTelemetryTriggerState == ESC_TLM_TRIGGER_PENDING)
+//     {
+//         return false;
+//     }
+//
+//     if (escTelemetryTriggerState == ESC_TLM_TRIGGER_READY)
+//     {
+//         if (debugMode == DEBUG_ESC_TELEMETRY) debug[0] = escTelemetryMotor+1;
+//
+//         if (escTelemetryMotor == index) {
+//             escTelemetryTriggerState = ESC_TLM_TRIGGER_PENDING;
+//             return true;
+//         }
+//     }
+//
+//     return false;
+// }
+
+bool escTelemetryInit(void)
+{
```

Comment:
Why not pass in the method (from pwm_output.h) to request telemetry here and store it?

the method only needs to take the uint8_t motorindex.